numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -1508,74 +1508,70 @@ PyUFunc_DefaultLegacyInnerLoopSelector(PyUFuncObject *ufunc,
+} _masked_stridedloop_data;
 
-    /* Allocate a new one */
-    n = (_ufunc_masker_data *)PyArray_malloc(sizeof(_ufunc_masker_data));
-    if (n == NULL) {
-        return NULL;
-    }
-
-    /* Copy the data (unmasked data doesn't have object semantics) */
-    memcpy(n, data, sizeof(_ufunc_masker_data));
 
-    return (NpyAuxData *)n;
+static void
+_masked_stridedloop_data_free(NpyAuxData *auxdata)
+{
+    _masked_stridedloop_data *data = (_masked_stridedloop_data *)auxdata;
+    PyMem_Free(data);
 }
 
+
 /*
  * This function wraps a regular unmasked ufunc inner loop as a
  * masked ufunc inner loop, only calling the function for
  * elements where the mask is True.
  */
 static void
 unmasked_ufunc_loop_as_masked(
-             char **dataptrs, npy_intp *strides,
-             char *mask, npy_intp mask_stride,
-             npy_intp loopsize,
-             NpyAuxData *innerloopdata)
+        char **data, const npy_intp *dimensions,
+        const npy_intp *strides, void *_auxdata)
 {
-    _ufunc_masker_data *data;
-    int iargs, nargs;
-    PyUFuncGenericFunction unmasked_innerloop;
-    void *unmasked_innerloopdata;
-    npy_intp subloopsize;
-
-    /* Put the aux data into local variables */
-    data = (_ufunc_masker_data *)innerloopdata;
-    unmasked_innerloop = data->unmasked_innerloop;
-    unmasked_innerloopdata = data->unmasked_innerloopdata;
-    nargs = data->nargs;
+    _masked_stridedloop_data *auxdata = (_masked_stridedloop_data *)_auxdata;
+    int nargs = auxdata->nargs;
+    PyUFuncGenericFunction strided_loop = auxdata->unmasked_stridedloop;
+    void *innerloopdata = auxdata->innerloopdata;
+
+    char **dataptrs = auxdata->dataptrs;
+    memcpy(dataptrs, data, nargs * sizeof(char *));
+    char *mask = data[nargs];
+    npy_intp mask_stride = strides[nargs];
 
+    npy_intp N = dimensions[0];
     /* Process the data as runs of unmasked values */
     do {
+        ssize_t subloopsize;
+
         /* Skip masked values */
-        mask = npy_memchr(mask, 0, mask_stride, loopsize, &subloopsize, 1);
-        for (iargs = 0; iargs < nargs; ++iargs) {
-            dataptrs[iargs] += subloopsize * strides[iargs];
+        mask = npy_memchr(mask, 0, mask_stride, N, &subloopsize, 1);
+        for (int i = 0; i < nargs; i++) {
+            dataptrs[i] += subloopsize * strides[i];
         }
-        loopsize -= subloopsize;
-        /*
-         * Process unmasked values (assumes unmasked loop doesn't
-         * mess with the 'args' pointer values)
-         */
-        mask = npy_memchr(mask, 0, mask_stride, loopsize, &subloopsize, 0);
-        unmasked_innerloop(dataptrs, &subloopsize, strides,
-                                        unmasked_innerloopdata);
-        for (iargs = 0; iargs < nargs; ++iargs) {
-            dataptrs[iargs] += subloopsize * strides[iargs];
-        }
-        loopsize -= subloopsize;
-    } while (loopsize > 0);
+        N -= subloopsize;
+
+        /* Process unmasked values */
```

Comment:
Gosh, never knew about this `memchr` stuff; should have used it for the reductions too - https://github.com/numpy/numpy/blob/cbec2c8054ea6150490b9e72eb051848b79344d1/numpy/core/src/umath/ufunc_object.c#L2956-L2959

Though if I look at source code of the `glibc` version, I have *no* idea why one would roll one's own - that is one heck of an optimized function!

Reply:
The followup deletes the whole chunk from the reductions to re-use this.  It might make sense to re-add a broadcasted special case eventually though.  Your reduce code has such a special case, the normal ufunc code does not.

Reply:
Nice! Quite happy to go for simpler first and then see whether speed can be improved for special cases.